IRQ Sharing, Hardware Timers & PCI devices

supercomputer7 opened the pull request:
This is a draft for now, so we can let travis to test all the changes.
Also, I might fix some parts after I recheck everything.

awesomekling commented on Kernel/Interrupts/InterruptManagement.h:
```diff
@@ -0,0 +1,67 @@
+
+protected:
+    InterruptManagement(bool create_default_controller);
+    FixedArray<IRQController*>& m_interrupt_controllers;
```
Instead of allocating member objects like this, using `new`, just declare them as regular members:

```suggestion
    FixedArray<IRQController*> { 1 };
```
```
```suggestion
    FixedArray<IRQController*>& m_interrupt_controllers;
```
```suggestion
    FixedArray<IRQController*>& m_interrupt_controllers;
```

supercomputer7 replied:
I fixed that, thanks! :)

supercomputer7 commented:
@awesomekling I still need to work on the commit messages. When I think it's ready, I'll let you know about it. For now, I let travis to check my patch again.

supercomputer7 commented:
@awesomekling Last recheck, then I think it's ready to be merged/reviewed :)

supercomputer7 commented:
Alright, it passed the last recheck :)

awesomekling commented on Kernel/Arch/i386/CPU.cpp:
```diff
@@ -444,16 +416,69 @@ static void unimp_trap()
+        if (s_interrupt_handler[interrupt_number]->purpose() == "Unhandled Interrupt Handler") {
+            s_interrupt_handler[interrupt_number] = &handler;
+            return;
+        } else if (s_interrupt_handler[interrupt_number]->is_shared_handler() && !s_interrupt_handler[interrupt_number]->is_sharing_with_others()) {
```
No need for `else` after `return`

awesomekling commented on Kernel/Arch/i386/CPU.cpp:
```diff
@@ -444,16 +416,69 @@ static void unimp_trap()
+            return;
+        } else if (s_interrupt_handler[interrupt_number]->is_shared_handler() && !s_interrupt_handler[interrupt_number]->is_sharing_with_others()) {
+            if (s_interrupt_handler[interrupt_number]->purpose() == "Shared IRQ Handler") {
+                reinterpret_cast<SharedIRQHandler*>(s_interrupt_handler[interrupt_number])->register_handler(handler);
```
Since `s_interrupt_handler` is an array of `GenericInterruptHandler*`, we can use `static_cast` here rather than `reinterpret_cast`

awesomekling commented on Kernel/Arch/i386/CPU.cpp:
```diff
@@ -444,16 +416,69 @@ static void unimp_trap()
+void register_generic_interrupt_handler(u8 interrupt_number, GenericInterruptHandler& handler)
+{
+    if (s_interrupt_handler[interrupt_number] != nullptr) {
+        if (s_interrupt_handler[interrupt_number]->purpose() == "Unhandled Interrupt Handler") {
```
Let's use an enum (or bool) instead of comparing strings in logic like this.

awesomekling commented on Kernel/init.cpp:
```diff
@@ -458,4 +461,31 @@ void setup_pci()
+    if (smp == "off") {
+        InterruptManagement::initialize();
+        return;
+    } else if (smp == "on") {
```
No need for `else` after `return`. Also, the same comment repeated everywhere else relevant. :)

awesomekling commented on Kernel/Interrupts/UnhandledInterruptHandler.cpp:
```diff
@@ -0,0 +1,50 @@
+
+#include <Kernel/Arch/i386/CPU.h>
+#include <Kernel/Interrupts/UnhandledInterruptHandler.h>
+namespace Kernel {
```
Please add an empty line after the last `#include` statement.

awesomekling commented on Kernel/Interrupts/UnhandledInterruptHandler.cpp:
```diff
@@ -0,0 +1,50 @@
+
+void UnhandledInterruptHandler::handle_interrupt(RegisterState&)
+{
+    dbgprintf("Interrupt: Unhandled vector %d was invoked for handle_interrupt(RegisterDump&).\n", interrupt_number());
```
Please prefer `dbg()` in new code.

awesomekling commented on Kernel/Interrupts/UnhandledInterruptHandler.h:
```diff
@@ -0,0 +1,51 @@
+#include <Kernel/Arch/i386/CPU.h>
+#include <Kernel/Interrupts/GenericInterruptHandler.h>
+namespace Kernel {
+class UnhandledInterruptHandler final : public GenericInterruptHandler {
```
Please add an empty line after the last `#include` statement.

awesomekling commented on Kernel/Interrupts/UnhandledInterruptHandler.h:
```diff
@@ -0,0 +1,51 @@
+class UnhandledInterruptHandler final : public GenericInterruptHandler {
+public:
+    explicit UnhandledInterruptHandler(u8 interrupt_vector);
+    ~UnhandledInterruptHandler();
```
For virtual classes, please specify `virtual` on the destructor.

awesomekling commented on Kernel/Interrupts/UnhandledInterruptHandler.h:
```diff
@@ -0,0 +1,51 @@
+
+    virtual String purpose() const override { return "Unhandled Interrupt Handler"; };
+
+    virtual size_t get_sharing_devices_count() const override { return 0; }
```
Since this not have an out-argument, let's call it `sharing_device_count()` or perhaps `number_of_sharing_devices()`.

supercomputer7 replied:
OK, I'll do so in a separate commit because I don't want to mess the patchset.
`sharing_device_count()` will be excellent :)

awesomekling commented on Kernel/Interrupts/UnhandledInterruptHandler.h:
```diff
@@ -0,0 +1,51 @@
+
+    virtual size_t get_sharing_devices_count() const override { return 0; }
+    virtual bool is_shared_handler() const override { return false; }
+    virtual bool is_sharing_with_others() const { return false; }
```
Please specify `override` whenever you are overriding a virtual member function.
This ensures that the compiler warns us if we rename the overridden function and forget to rename this as well :)

shannonbooth commented on Kernel/Devices/PIT.cpp:
```diff
@@ -0,0 +1,98 @@
+
+PIT& PIT::the()
+{
+    if (s_the == nullptr) {
```
Might be missing something here, but isn't s_the always set in initialize?

supercomputer7 replied:
Well, I can replace this with assertion. The idea is that if someone call the PIT singleton class before initializing it, it won't fail.

shannonbooth replied:
Yeah fair enough. I just saw similar code between the() and initialize() -- and the pattern looked slightly different to me from what I was used to for whatever reason :)

shannonbooth commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,229 @@
+void InterruptManagement::switch_to_ioapic_mode()
+{
+    kprintf("Interrupts: Switch to IOAPIC mode\n");
+    if (!initialized)
```
For these calls, should being initialized be a precondition? Seems a little strange to me for these calls to silently fail.

supercomputer7 replied:
I should replace it with assertion, thank you! :)

supercomputer7 commented on Kernel/Arch/i386/CPU.cpp:
```diff
@@ -444,16 +416,69 @@ static void unimp_trap()
+        return;
+    }
+    if (s_interrupt_handler[interrupt_number]->is_shared_handler() && !s_interrupt_handler[interrupt_number]->is_sharing_with_others()) {
+        if (s_interrupt_handler[interrupt_number]->purpose() == "Shared IRQ Handler") {
```
That's a good idea :)

supercomputer7 commented:
@awesomekling @shannonbooth thanks for the reviews :)
I'm re-checking the code now, found some nitpicks to fix.

supercomputer7 commented on Kernel/Interrupts/PIC.h:
```diff
@@ -0,0 +1,52 @@
+    virtual u16 get_isr() const override;
+    virtual u16 get_irr() const override;
+    virtual u32 get_gsi_base() const override { return 0; }
+    virtual String get_model() const override { return "Dual Intel 8259"; }
```
I'm using this method to determine the IRQ controller type but I don't think that's a good idea.
@awesomekling what do you think about it? Maybe I should write an enum?

supercomputer7 replied:
I decided to write another method called type() that will return the type of the IRQController from an enum.

supercomputer7 commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+
+    for (auto* irq_controller : InterruptManagement::the().m_interrupt_controllers) {
+        ASSERT(irq_controller != nullptr);
+        if (irq_controller->get_model().matches("Dual Intel 8259")) {
```
Now that I see that, I think I should change it, and I shouldn't use String::matches() method here anyway...

awesomekling commented on Kernel/Interrupts/IRQController.h:
```diff
@@ -0,0 +1,64 @@
+protected:
+    IRQController()
+        : m_enabled(false)
+        , m_hard_disabled(false)
```
Please initialize these at the declarations below instead of using a constructor initializer list.

supercomputer7 commented on Kernel/ACPI/MultiProcessorParser.cpp:
```diff
@@ -0,0 +1,268 @@
+    char* p_floating_pointer_str = (char*)(PhysicalAddress(ebda_segment << 4).as_ptr());
+    for (char* floating_pointer_str = (char*)floating_pointer_region->vaddr().offset(offset_in_page((u32)(ebda_segment << 4))).as_ptr(); floating_pointer_str < (char*)(floating_pointer_region->vaddr().offset(offset_in_page((u32)(ebda_segment << 4))).get() + 1024); floating_pointer_str += 16) {
+#ifdef MUTLIPROCESSOR_DEBUG
+        dbgprintf("MultiProcessor: Looking for floating pointer structure in EBDA @ V0x%x, P0x%x\n", floating_pointer_str, p_floating_pointer_str);
```
I should replace `dbgprintf()` with `dbg()` in this file.

awesomekling commented on Kernel/ACPI/MultiProcessorParser.cpp:
```diff
@@ -0,0 +1,268 @@
+
+Vector<unsigned>& MultiProcessorParser::get_pci_bus_ids()
+{
+    Vector<unsigned>& pci_bus_ids = *new Vector<unsigned>();
```
Please don't abuse `operator new` and references like this.
References and raw pointers should only refer to objects kept alive by *someone else*.
Also, as there are no corresponding `delete` statements, objects allocated this way will leak.

Instead, simply use value semantics by returning a `Vector<unsigned>` and creating a `Vector<unsigned> on the stack:

```cpp
Vector<unsigned> MultiProcessorParser::get_pci_bus_ids()
{
    Vector<unsigned> pci_bus_ids;
    ....
    return pci_bus_ids;
}
```

awesomekling commented on Kernel/ACPI/MultiProcessorParser.cpp:
```diff
@@ -0,0 +1,268 @@
+
+        auto entry_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)entry)), PAGE_ROUND_UP(m_configuration_table_length), "MultiProcessor Parser Parsing Bus Entry", Region::Access::Read, false, true);
+        auto* v_entry_ptr = (MultiProcessor::IOInterruptAssignmentEntry*)entry_region->vaddr().offset(offset_in_page((u32)entry)).as_ptr();
+        dbg() << "MultProcessor: Parsing Entry P 0x" << String::format("%x", entry) << ", V 0x" << String::format("%x", v_entry_ptr);
```
I think you can write this as:
```suggestion
        dbg() << "MultiProcessor: Parsing Entry P " << entry << ", V 0x" << v_entry_ptr;
```

supercomputer7 replied:
I'll fix the typo and remove the String::format() since we don't need it, but shouldn't we output addresses with hexadecimal notation (e.g. using "0x" before the number) ?

supercomputer7 replied:
My mistake, I see that the PrintfImplementation with `dbg()` already prints out hexadecimal prefix :)

supercomputer7 replied:
@awesomekling The `entry` variable isn't pointer, therefore I still need to call `String::format()` on it, but `v_entry_ptr` is indeed a pointer so no need to call `String::format()` for it.

awesomekling commented on Kernel/ACPI/MultiProcessorParser.h:
```diff
@@ -0,0 +1,241 @@
+
+}
+
+class PCIInterruptOverrideMetadata : public RefCounted<PCIInterruptOverrideMetadata> {
```
This class doesn't seem like it needs reference counting. Is it expected to have multiple owners that keep it alive? Is it expected to be mutable?

supercomputer7 replied:
It's expected to have multiple owners - the Interrupt management will hold a reference and the corresponding IOAPIC object which handles the overriding.

awesomekling commented on Kernel/ACPI/MultiProcessorParser.h:
```diff
@@ -0,0 +1,241 @@
+    u8 pci_interrupt_pin();
+    u8 pci_device_number();
+    u32 ioapic_id();
+    u16 ioapic_interrupt_pin();
```
These should be `const`. Also they can be inline.

awesomekling commented on Kernel/ACPI/MultiProcessorParser.h:
```diff
@@ -0,0 +1,241 @@
+    MultiProcessor::FloatingPointer* m_floating_pointer;
+    MultiProcessor::ConfigurationTableHeader* m_configuration_table;
+    Vector<MultiProcessor::IOInterruptAssignmentEntry*> m_io_interrupt_redirection_entries;
+    Vector<MultiProcessor::BusEntry*> m_bus_entries;
```
If I understand correctly, these are all offsets, not actually pointers.
As such, we should not store them as pointers, as it's not safe to dereference them.

awesomekling commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+    if (!initialized)
+        s_interrupt_management = new InterruptManagement(true);
+    initialized = true;
+}
```
It's very hard to understand the lifetime model of this kind of object.
Instead, I suggest getting rid of automatic instantiation and creating objects like this explicitly on the stack in `init.cpp`
That way, we can see what objects are being created during boot and when.

supercomputer7 replied:
I'll work on a solution, I agree it's pretty much complicated to see how it is initialized and created.

awesomekling commented on Kernel/Interrupts/IRQController.h:
```diff
@@ -0,0 +1,64 @@
+    virtual u16 get_isr() const = 0;
+    virtual u16 get_irr() const = 0;
+    virtual String get_model() const = 0;
+    virtual u8 type() const = 0;
```
This should return `IRQControllerType`.
Then you don't have to cast to `u8` every time when comparing :)

awesomekling commented on Kernel/Interrupts/IRQController.h:
```diff
@@ -0,0 +1,64 @@
+    virtual u32 get_gsi_base() const = 0;
+    virtual u16 get_isr() const = 0;
+    virtual u16 get_irr() const = 0;
+    virtual String get_model() const = 0;
```
This should return `const char*`, then you don't create a heap allocation every time you call `get_model()` :)

supercomputer7 replied:
I should probably change the name of this method.

awesomekling commented on Kernel/Interrupts/IRQController.h:
```diff
@@ -0,0 +1,64 @@
+    virtual void disable(u8 number) = 0;
+    virtual void hard_disable() { m_hard_disabled = true; }
+    virtual bool is_vector_enabled(u8 number) const = 0;
+    virtual bool is_enabled() const { return m_enabled && !m_hard_disabled; }
```
Should this be `virtual`? Are subclasses going to override it?

supercomputer7 replied:
`is_enabled()` and `is_hard_disabled` can be non-virtual methods, but `hard_disable()` must be virtual because hard-disabling the PIC/IOAPIC should be implemented accordingly.

awesomekling commented on Kernel/Interrupts/GenericInterruptHandler.h:
```diff
@@ -0,0 +1,74 @@
+    virtual bool is_shared_handler() const = 0;
+    virtual bool is_sharing_with_others() const = 0;
+
+    virtual u8 purpose() const = 0;
```
This should return `HandlerPurpose`.

awesomekling commented on Kernel/Interrupts/IRQHandler.cpp:
```diff
@@ -0,0 +1,86 @@
+IRQHandler::IRQHandler(u8 irq)
+    : GenericInterruptHandler(irq)
+    , m_shared_with_others(false)
+    , m_enabled(false)
```
Please initialize members when declaring them whenever possible instead of in a constructor initializer list.

awesomekling commented on Kernel/Interrupts/InterruptManagement.h:
```diff
@@ -0,0 +1,65 @@
+#include <Kernel/Interrupts/GenericInterruptHandler.h>
+#include <Kernel/Interrupts/IOAPIC.h>
+#include <Kernel/Interrupts/IRQController.h>
+namespace Kernel {
```
Please leave some space after the last `#include`.

awesomekling commented on Kernel/Interrupts/InterruptManagement.h:
```diff
@@ -0,0 +1,65 @@
+    FixedArray<IRQController*> m_interrupt_controllers { 1 };
+};
+
+class ISAInterruptOverrideMetadata : public RefCounted<ISAInterruptOverrideMetadata> {
```
Does this class really need to be reference-counted?

supercomputer7 replied:
It's expected to have multiple owners - the Interrupt management will hold a reference and the corresponding IOAPIC object which handles the overriding.

awesomekling commented on Kernel/Interrupts/InterruptManagement.h:
```diff
@@ -0,0 +1,65 @@
+    u8 bus();
+    u8 source();
+    u32 gsi();
+    u16 flags();
```
These should all be `const` and could also be inline.

awesomekling commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+
+InterruptManagement::InterruptManagement(bool create_default_controller)
+{
+    m_interrupt_controllers.resize(1);
```
This is already constructed with size 1 per the initializer at the declaration. No need to resize it again.

supercomputer7 commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+    , m_madt(p_madt)
+{
+    // FIXME: Check what is the actual data size then map accordingly
+    dbgprintf("Interrupts: MADT @ P0x%x\n", &p_madt);
```
`dbgprintf()` to `dbg()` .

awesomekling commented on Kernel/Interrupts/InterruptManagement.h:
```diff
@@ -0,0 +1,65 @@
+
+protected:
+    InterruptManagement(bool create_default_controller);
+    FixedArray<IRQController*> m_interrupt_controllers { 1 };
```
Please remove all unnecessary use of `new`.
If you make this a `FixedArray<NonnullOwnPtr<IRQController>>`, you can create new entries for it using `make<IRQController>(...)` and they will not leak.

supercomputer7 commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+        ASSERT(irq_controller != nullptr);
+        if (irq_controller->type() == (u8)IRQControllerType::i8259) {
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
```
`dbgprintf()` to `dbg()` .

supercomputer7 commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
+        } else {
+            dbgprintf("Interrupts: Detected %s\n", irq_controller->get_model().characters());
```
`dbgprintf()` to `dbg()` .

supercomputer7 commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+        ASSERT(irq_controller != nullptr);
+        if (irq_controller->type() == (u8)IRQControllerType::i82093AA) {
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
```
`dbgprintf()` to `dbg()` .

supercomputer7 commented on Kernel/Interrupts/InterruptManagement.cpp:
```diff
@@ -0,0 +1,227 @@
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
+        } else {
+            dbgprintf("Interrupts: Detected %s\n", irq_controller->get_model().characters());
```
`dbgprintf()` to `dbg()` .

supercomputer7 commented:
@awesomekling I added 2 commits for fixing some "critical" ACPI stuff, that is still related to the PR.
You commented about that in `MultiProcessorParser` class we shouldn't use pointers as offsets (or in other words - physical addresses), since it can lead to accidental dereference. Therefore I changed in the ACPI code (and also in the PCI MMIO class) to use the `PhysicalAddress` class instead of raw pointers.

supercomputer7 commented on Kernel/ACPI/ACPIParser.cpp:
```diff
@@ -58,10 +58,10 @@ ACPIParser::ACPIParser(bool usable)
 {
     kprintf("ACPI: Requested to search for a table, Abort!\n");
-    return nullptr;
+    return PhysicalAddress((uintptr_t) nullptr);
```
Is calling `PhysicalAddress()` constructor will create a `PhysicalAddress` that is equivalent to `PhysicalAddress((uintptr_t) nullptr)`?

supercomputer7 replied:
thanks to @Quaker762 that explained to me in the IRC channel that I can use the default constructor, I'll fix it :)

supercomputer7 commented on Kernel/ACPI/ACPIStaticParser.h:
```diff
@@ -33,34 +33,34 @@ namespace Kernel {
+    PhysicalAddress search_rsdp_in_bios_area();
+    PhysicalAddress search_rsdp();
 
     // Early pointers that are needed really for initializtion only...
```
Oops, wrong comment here...

Quaker762 commented on Kernel/ACPI/ACPIParser.cpp:
```diff
@@ -58,10 +58,10 @@ ACPIParser::ACPIParser(bool usable)
 {
     kprintf("ACPI: Requested to search for a table, Abort!\n");
-    return nullptr;
+    return PhysicalAddress((uintptr_t) nullptr);
```
It's pretty janky to pass `nullptr` to the constructor of `PhysicalAddress`. As you mentioned in IRC, change this to `return {};`

https://stackoverflow.com/questions/39487065/what-does-return-statement-mean-in-c11

supercomputer7 replied:
Sure will do, thanks :)

shannonbooth commented on Kernel/Interrupts/IOAPIC.h:
```diff
@@ -0,0 +1,87 @@
+    void unmask_redirection_entry(u8 index);
+    bool is_redirection_entry_masked(u8 index) const;
+
+    u8 read_redirection_entry_vector(u8 index);
```
I think these can be made const

awesomekling commented on Kernel/ACPI/ACPIStaticParser.h:
```diff
@@ -61,8 +61,8 @@ class ACPIStaticParser : ACPIParser {
-    OwnPtr<ACPI::MainSystemDescriptionTable> m_main_sdt;
-    OwnPtr<ACPI::FixedACPIData> m_fadt;
+    OwnPtr<ACPI::MainSystemDescriptionTable> m_main_sdt { nullptr };
+    OwnPtr<ACPI::FixedACPIData> m_fadt { nullptr };
```
No need to explicitly initialize these. `OwnPtr` automatically sets itself to `nullptr`.

awesomekling commented:
Very nice improvements. The use of `PhysicalAddress` turned out really nice :)
